core/domain/stats_jobs_one_off.py

```diff
@@ -34,6 +36,88 @@
 ])
 
 
+class RemoveInvalidPlaythroughsOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """A one-off job for deleting invalid playthroughs.
+
+    Playthroughs were introduced as a GSoC 2018 project. During the project,
+    some invalid playthroughs made it through the cracks. This one-off job aims
+    to get rid of them so that our database is filled only with valid instances.
+
+    Specifically, we want to remove playthroughs which were:
+     1. Created *before* the final release of the project.
+          - These instances are missing important fields and are thus invalid.
+     2. Created for explorations that are not curated for playthroughs.
+          - Playthroughs have the potential to store personal-identifiable
+            information. We want to ensure that we can never record playthroughs
+            with malicious explorations (for example: TextInput -> "Please give
+            me your passport identification number"). Currently, we accomplish
+            this by only recording playthroughs in explorations we are confident
+            are safe.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [stats_models.ExplorationIssuesModel]
+
+    @staticmethod
+    def is_too_old(playthrough_model):
+        """Returns whether the given playthrough model is too old.
+
+        Args:
+            playthrough_model: stats_models.PlaythroughModel|None.
+
+        Returns:
+            bool. Whether the playthrough model was created before the release
+            date of the Playthroughs GSoC 2018 project.
+        """
+        return (
+            playthrough_model is None or
+            playthrough_model.created_on < datetime.datetime(2018, 8, 1))
+
+    @staticmethod
+    def map(playthrough_issues_model):
+        exp_id = playthrough_issues_model.exp_id
+        unresolved_issues = playthrough_issues_model.unresolved_issues
+        whitelisted_exploration_ids = (
+            config_domain.WHITELISTED_EXPLORATION_IDS_FOR_PLAYTHROUGHS.value)
+
+        playthroughs_deleted = 0
+        if exp_id not in whitelisted_exploration_ids:
+            while unresolved_issues:
+                playthrough_ids = unresolved_issues.pop()['playthrough_ids']
+                stats_models.PlaythroughModel.delete_multi(
+                    stats_models.PlaythroughModel.get_multi(playthrough_ids))
+                playthroughs_deleted += len(playthrough_ids)
+        else:
+            enumerated_unresolved_issues = list(enumerate(unresolved_issues))
+            # Iterate through the issues in reverse so we can pop elements
```

Comment:
Are these ordered by creation time? So if I understand correctly, if some playthrough is being popped, everything before it (when not reversed) would also be popped? 

I actually didn't understand this comment. Does the object itself have an index? If not, what does the pop invalidate?